File: src/ui/lcd/gate-edit.ts
/**
 * LCD Gate Edit screen — gate pattern with step toggle via 16 buttons.
 * 2×8 grid layout matching the physical step button grid.
 * All text ≥16px for readability on 3.5" TFT at 50cm.
 */

import type { SequencerState } from '../../engine/types'
import { COLORS } from '../colors'
import type { UIState } from '../hw-types'
import { drawText, fillRect, LCD_CONTENT_H, LCD_CONTENT_Y, LCD_W, strokeRect } from '../renderer'

const PAD = 8
const HEADER_H = 42
const STEP_AREA_TOP = LCD_CONTENT_Y + HEADER_H
const COLS = 8
const ROW_GAP = 8
const COL_GAP = 2
const STEP_W = (LCD_W - PAD * 2 - (COLS - 1) * COL_GAP) / COLS
const AVAIL_H = LCD_CONTENT_H - HEADER_H - 4
const STEP_H = (AVAIL_H - ROW_GAP) / 2

export function renderGateEdit(ctx: CanvasRenderingContext2D, engine: SequencerState, ui: UIState): void {
  const track = engine.tracks[ui.selectedTrack]
  const trackColor = COLORS.track[ui.selectedTrack]
  const pageOffset = ui.currentPage * 16
  const maxPage = Math.max(0, Math.ceil(track.gate.length / 16) - 1)

  const selectedStepIdx = ui.selectedStep >= 0 ? ui.currentPage * 16 + ui.selectedStep : -1

  // Title
  drawText(ctx, `GATE — T${ui.selectedTrack + 1}`, PAD, LCD_CONTENT_Y + 18, trackColor, 18)

  // Right side info: show gate length + ratchet when a step is selected
  let infoText = ''
  if (selectedStepIdx >= 0 && selectedStepIdx < track.gate.length) {
    const step = track.gate.steps[selectedStepIdx]
    const gl = Math.round(step.length * 100)
    const rc = step.ratchet
    infoText = `GL:${gl}%`
    if (rc > 1) infoText += ` R:${rc}x`
  }
  if (track.gate.clockDivider > 1) infoText += `  ÷${track.gate.clockDivider}`
  if (maxPage > 0) infoText += `  P${ui.currentPage + 1}/${maxPage + 1}`
  if (!infoText) infoText = `LEN ${track.gate.length}`
  drawText(ctx, infoText, LCD_W - PAD, LCD_CONTENT_Y + 18, COLORS.textDim, 16, 'right')

  // Step grid — 2 rows of 8
  for (let row = 0; row < 2; row++) {
    const rowY = STEP_AREA_TOP + row * (STEP_H + ROW_GAP)
    for (let col = 0; col < COLS; col++) {
      const i = row * COLS + col
      const stepIdx = pageOffset + i
      const x = PAD + col * (STEP_W + COL_GAP)

      if (stepIdx >= track.gate.length) {
        fillRect(ctx, { x, y: rowY, w: STEP_W, h: STEP_H }, '#111118')
      } else if (track.gate.steps[stepIdx].tie) {
        // Tied step: filled with track color at reduced opacity
        fillRect(ctx, { x, y: rowY, w: STEP_W, h: STEP_H }, COLORS.trackDim[ui.selectedTrack])
        fillRect(ctx, { x, y: rowY, w: STEP_W, h: STEP_H }, `${trackColor}88`)

        // Draw connecting bridge from previous step
        if (col > 0) {
          // Horizontal bridge across the gap between this step and the previous
          const bridgeY = rowY + Math.round(STEP_H / 2) - 2
          fillRect(ctx, { x: x - COL_GAP, y: bridgeY, w: COL_GAP, h: 4 }, trackColor)
        } else if (row > 0) {
          // Row wrap: draw indicator at left edge of this row and right edge of previous row
          const bridgeY = rowY + Math.round(STEP_H / 2) - 2
          fillRect(ctx, { x: PAD, y: bridgeY, w: 3, h: 4 }, trackColor)
          // Previous row right edge indicator
          const prevRowY = STEP_AREA_TOP + (row - 1) * (STEP_H + ROW_GAP)
          const prevRowRight = PAD + (COLS - 1) * (STEP_W + COL_GAP) + STEP_W
          fillRect(ctx, { x: prevRowRight - 3, y: prevRowY + Math.round(STEP_H / 2) - 2, w: 3, h: 4 }, trackColor)
        }
      } else if (track.gate.steps[stepIdx].on) {
        // Gate ON: show gate length as filled portion of step cell
        const step = track.gate.steps[stepIdx]
        const barH = Math.round(STEP_H * step.length)
        fillRect(ctx, { x, y: rowY, w: STEP_W, h: STEP_H }, COLORS.trackDim[ui.selectedTrack])
        fillRect(ctx, { x, y: rowY + (STEP_H - barH), w: STEP_W, h: barH }, trackColor)

        // Ratchet tick marks — horizontal lines dividing the bar
        if (step.ratchet > 1) {
          const rc = step.ratchet
          for (let r = 1; r < rc; r++) {
            const tickY = rowY + (STEP_H - barH) + Math.round((barH / rc) * r)
            fillRect(ctx, { x: x + 2, y: tickY, w: STEP_W - 4, h: 1 }, '#000000aa')
          }
        }

        // Draw bridge to next step if next is tied
        if (stepIdx + 1 < track.gate.length && track.gate.steps[stepIdx + 1].tie) {
          if (col < COLS - 1) {
            const bridgeY = rowY + Math.round(STEP_H / 2) - 2
            fillRect(ctx, { x: x + STEP_W, y: bridgeY, w: COL_GAP, h: 4 }, trackColor)
          }
        }
      } else {
        fillRect(ctx, { x, y: rowY, w: STEP_W, h: STEP_H }, COLORS.trackDim[ui.selectedTrack])
      }

      // Selected step highlight
      if (i === ui.selectedStep && ui.selectedStep >= 0) {
        strokeRect(ctx, { x: x - 1, y: rowY - 1, w: STEP_W + 2, h: STEP_H + 2 }, '#ffffff', 2)
      }
      // Playhead indicator
      else if (stepIdx === track.gate.currentStep) {
        strokeRect(ctx, { x: x - 1, y: rowY - 1, w: STEP_W + 2, h: STEP_H + 2 }, '#ffffff88', 1)
      }
    }
  }
}

File: crates/renderer/src/screens/gate_edit.rs
use embedded_graphics::{pixelcolor::Rgb565, prelude::*};
use requencer_engine::types::SequencerState;

use crate::{colors, draw, layout, types::UiState};

/// Render the gate edit screen: 2x8 step grid showing gate on/off, length, ratchet, tie.
pub fn render<D: DrawTarget<Color = Rgb565>>(
    display: &mut D,
    state: &SequencerState,
    ui: &UiState,
) {
    let track_idx = ui.selected_track as usize;
    let track = &state.tracks[track_idx];
    let color = colors::TRACK[track_idx];
    let dim = colors::TRACK_DIM[track_idx];
    let page_offset = ui.current_page as usize * 16;
    let (cell_w, cell_h) = layout::edit_step_size();

    // Header
    draw::fill_rect(
        display,
        0,
        layout::CONTENT_Y as i32,
        layout::LCD_W,
        layout::EDIT_HEADER_H,
        colors::STATUS_BAR,
    );
    draw::text(
        display,
        layout::PAD as i32,
        layout::CONTENT_Y as i32 + 9,
        "GATE",
        color,
    );

    // Length / divider info
    let mut buf = [0u8; 16];
    let len_str = draw::format_u16(track.gate.length as u16, &mut buf);
    draw::text_right(
        display,
        layout::LCD_W as i32 - layout::PAD as i32,
        layout::CONTENT_Y as i32 + 9,
        len_str,
        colors::TEXT_DIM,
    );

    // Step grid: 2 rows x 8 cols
    let grid_y = layout::CONTENT_Y as i32 + layout::EDIT_HEADER_H as i32;

    for row in 0..2u32 {
        for col in 0..8u32 {
            let step_idx = page_offset + (row * 8 + col) as usize;
            if step_idx >= track.gate.steps.len() {
                continue;
            }
            let step = &track.gate.steps[step_idx];
            let x = layout::PAD as i32 + col as i32 * cell_w as i32;
            let y = grid_y + row as i32 * cell_h as i32;
            let is_playhead = track.gate.current_step as usize == step_idx;
            let is_selected = ui.selected_step as usize == step_idx;

            // Cell background
            let bg = if is_playhead {
                colors::SELECTED_ROW
            } else {
                colors::LCD_BG
            };
            draw::fill_rect(display, x + 1, y + 1, cell_w - 2, cell_h - 2, bg);

            // Gate state visualization
            if step.on {
                // Filled portion represents gate length
                let fill_h = ((step.length * (cell_h - 4) as f32) as u32).max(2);
                let fill_y = y + (cell_h as i32 - fill_h as i32 - 2);
                draw::fill_rect(display, x + 2, fill_y, cell_w - 4, fill_h, color);

                // Ratchet indicator (dots at top)
                if step.ratchet > 1 {
                    for r in 0..step.ratchet.min(4) {
                        let dot_x = x + 3 + r as i32 * 4;
                        draw::fill_rect(display, dot_x, y + 2, 2, 2, colors::TEXT_BRIGHT);
                    }
                }
            } else if step.tie {
                // Tie: horizontal bridge
                draw::fill_rect(
                    display,
                    x,
                    y + cell_h as i32 / 2 - 2,
                    cell_w,
                    4,
                    dim,
                );
            }

            // Selection outline
            if is_selected {
                draw::stroke_rect(display, x, y, cell_w, cell_h, colors::TEXT_BRIGHT);
            }
        }
    }

    // Footer: page info
    let footer_y = layout::LCD_H as i32 - layout::EDIT_FOOTER_H as i32;
    draw::fill_rect(
        display,
        0,
        footer_y,
        layout::LCD_W,
        layout::EDIT_FOOTER_H,
        colors::STATUS_BAR,
    );
    let mut page_buf = [0u8; 16];
    let page_str = draw::format_u16(ui.current_page as u16 + 1, &mut page_buf);
    draw::text_center(
        display,
        layout::LCD_W as i32 / 2,
        footer_y + 6,
        page_str,
        colors::TEXT_DIM,
    );
}
